pandas/tests/test_frame.py

```diff
@@ -12385,6 +12404,93 @@ def verify(df):
                 for col in ['4th', '5th']:
                     verify(udf[col])
 
+        # GH7403
+        df = pd.DataFrame({'A': list('aaaabbbb'),'B':range(8), 'C':range(8)})
+        df.iloc[3, 1] = np.NaN
+        left = df.set_index(['A', 'B']).unstack(0)
+
+        vals = [[3, 0, 1, 2, nan, nan, nan, nan],
+                [nan, nan, nan, nan, 4, 5, 6, 7]]
+        vals = list(map(list, zip(*vals)))
+        idx = Index([nan, 0, 1, 2, 4, 5, 6, 7], name='B')
```

Comment:
It may not be obvious (at least it wasn't to me), but ``idx = Index([nan, 0, 1, 2, 4, 5, 6, 7], name='B')`` results in ``Float64Index([nan, 0.0, 1.0, 2.0, 4.0, 5.0, 6.0, 7.0], dtype='float64')`` -- i.e. the integer *labels* get converted to floats. Thus you are testing that when unstacking a MultiIndex that leaves a single level with integer labels and ``NaN``s, the result is a ``Float64Index``. This is pretty much unavoidable -- unless you explicitly cast the integers to ``object``s rather than ``float64`` -- as MultIndex doesn't seem to support a single level. (Note that an integer  MultiIndex level can have missing values, since they are represented by ``-1`` in the labels and not ``NaN`` in the labels.) I think either (a) MultIndex should support a single level (and not convert to an Index), or (b) the unstacking code should convert such an integer-with-NaN index to ``object`` rather than ``float64``. [I encountered this issue while trying to make the ``stack()`` code handle ``NaN`` labels propertly.]